File: app/src/main/java/io/github/mohamedisoliman/androiddev/data/remote/RedditRemoteStore.java
package io.github.mohamedisoliman.androiddev.data.remote;

import io.github.mohamedisoliman.androiddev.data.model.Child;
import io.github.mohamedisoliman.androiddev.data.model.Data;
import io.github.mohamedisoliman.androiddev.data.model.RedditPost;
import io.github.mohamedisoliman.androiddev.data.model.RedditResponse;
import io.reactivex.Observable;
import java.util.List;

/**
 * Created by Mohamed Ibrahim on 8/1/18.
 */
public class RedditRemoteStore {

  private static final String ANDROIDDEV_SUBREDDIT = "androiddev";
  private static final int LIMIT = 10;

  private RedditApi redditApi;

  public RedditRemoteStore(RedditApi redditApi) {
    this.redditApi = redditApi;
  }

  public Observable<List<RedditPost>> getSubreddit() {
    return redditApi.getSubreddit(ANDROIDDEV_SUBREDDIT, "", LIMIT)
        .map(RedditResponse::getData)
        .map(Data::getChildren)
        .flatMap(Observable::fromIterable)
        .map(Child::getData)
        .map(data -> new RedditPost(data.getId(), data.getTitle(), data.getAuthor(), data.getUrl(),
            data.getThumbnail(), data.getUps(), data.getCreated()))
        .toList()
        .toObservable();
  }
}

File: app/src/main/java/io/github/mohamedisoliman/androiddev/data/remote/RedditRemoteStore.kt
package io.github.mohamedisoliman.androiddev.data.remote

import io.github.mohamedisoliman.androiddev.data.model.RedditPost
import io.reactivex.Observable
import timber.log.Timber

/**
 * Created by Mohamed Ibrahim on 8/1/18.
 */
class RedditRemoteStore(private val redditApi: RedditApi) {

  companion object {
    private const val ANDROIDDEV_SUBREDDIT = "androiddev"
  }

  fun getSubreddit(filter: String, limit: Int): Observable<List<RedditPost>> {
    return redditApi.getSubreddit(ANDROIDDEV_SUBREDDIT, filter, limit = limit)
        .doOnError { Timber.e(it) }
        .map { redditResponse -> redditResponse.data }
        .map { it.children }
        .flatMap { source -> Observable.fromIterable(source) }
        .map { child -> child.data }
        .map { RedditPost(0, it.id, it.title, it.author, it.url, it.thumbnail, it.ups, it.created) }
        .toList()
        .toObservable()
  }

}
